<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>CoMM: a collaborative mixed model to dissecting genetic contributions to complex traits by leveraging regulatory information • CoMM</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="CoMM: a collaborative mixed model to dissecting genetic contributions to complex traits by leveraging regulatory information">
<meta property="og:description" content="CoMM">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CoMM</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/CoMM.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/gordonliu810822/CoMM/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="CoMM_files/accessible-code-block-0.0.1/empty-anchor.js"></script><link href="CoMM_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="CoMM_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>CoMM: a collaborative mixed model to dissecting genetic contributions to complex traits by leveraging regulatory information</h1>
                        <h4 class="author">Jin Liu</h4>
            
            <h4 class="date">2021-08-31</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/gordonliu810822/CoMM/blob/master/vignettes/CoMM.Rmd"><code>vignettes/CoMM.Rmd</code></a></small>
      <div class="hidden name"><code>CoMM.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level3">
<h3 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h3>
<p>This vignette provides an introduction to the <code>CoMM</code> package. R package <code>CoMM</code> implements CoMM, a collaborative mixed model to dissecting genetic contributions to complex traits by leveraging regulatory information. The package can be installed with the command:</p>
<p><code><a href="https://devtools.r-lib.org/">library(devtools)</a></code></p>
<p><code>install_github("gordonliu810822/CoMM")</code></p>
<p>The package can be loaded with the command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/gordonliu810822/CoMM">"CoMM"</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="fit-comm-using-simulated-data" class="section level3">
<h3 class="hasAnchor">
<a href="#fit-comm-using-simulated-data" class="anchor"></a>Fit CoMM using simulated data</h3>
<p>We first generate genotype data using function <em>genRawGeno</em>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://mvtnorm.R-forge.R-project.org">mvtnorm</a></span><span class="op">)</span>
<span class="va">L</span> <span class="op">=</span> <span class="fl">1</span>; <span class="va">M</span> <span class="op">=</span> <span class="fl">100</span>; <span class="va">rho</span> <span class="op">=</span><span class="fl">0.5</span>
<span class="va">n1</span> <span class="op">=</span> <span class="fl">350</span>; <span class="va">n2</span> <span class="op">=</span> <span class="fl">5000</span>;
<span class="va">maf</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="va">M</span>,<span class="fl">0.05</span>,<span class="fl">0.5</span><span class="op">)</span>
<span class="va">X</span> <span class="op">=</span> <span class="fu">genRawGeno</span><span class="op">(</span><span class="va">maf</span>, <span class="va">L</span>, <span class="va">M</span>, <span class="va">rho</span>, <span class="va">n1</span> <span class="op">+</span> <span class="va">n2</span><span class="op">)</span>;</code></pre></div>
<p>Then, effect sizes are generated from standard Gaussian distribution with sparse structure:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">beta_prop</span> <span class="op">=</span> <span class="fl">0.2</span>;
<span class="va">b</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">numeric</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span>;
<span class="va">m</span> <span class="op">=</span> <span class="va">M</span> <span class="op">*</span> <span class="va">beta_prop</span>;
<span class="va">b</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">m</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span>;</code></pre></div>
<p>Subsequently, the gene expression <code>y</code> is generated by controlling cellular heritability at prespecified level (<code>h2y</code>):</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">h2y</span> <span class="op">=</span> <span class="fl">0.05</span>;
<span class="va">b0</span> <span class="op">=</span> <span class="fl">6</span>;
<span class="va">y0</span> <span class="op">&lt;-</span> <span class="va">X</span><span class="op">%*%</span><span class="va">b</span> <span class="op">+</span> <span class="va">b0</span>;
<span class="va">y</span>  <span class="op">&lt;-</span> <span class="va">y0</span> <span class="op">+</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">var</a></span><span class="op">(</span><span class="va">y0</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">h2y</span><span class="op">)</span><span class="op">/</span><span class="va">h2y</span><span class="op">)</span><span class="op">)</span><span class="op">^</span><span class="fl">0.5</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">n1</span><span class="op">+</span><span class="va">n2</span><span class="op">)</span>;</code></pre></div>
<p>Finally, the phenotype data is generated as the generative model of CoMM with a prespecified trait heritability (<code>h2</code>) as:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">h2</span> <span class="op">=</span> <span class="fl">0.001</span>;
<span class="va">y1</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n1</span><span class="op">]</span>
<span class="va">X1</span> <span class="op">&lt;-</span> <span class="va">X</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n1</span>,<span class="op">]</span>
<span class="va">y2</span> <span class="op">&lt;-</span> <span class="va">y0</span><span class="op">[</span><span class="op">(</span><span class="va">n1</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="va">n1</span><span class="op">+</span><span class="va">n2</span><span class="op">)</span><span class="op">]</span>
<span class="va">X2</span> <span class="op">&lt;-</span> <span class="va">X</span><span class="op">[</span><span class="op">(</span><span class="va">n1</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="va">n1</span><span class="op">+</span><span class="va">n2</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">alpha0</span> <span class="op">&lt;-</span> <span class="fl">3</span>  
<span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.3</span>
<span class="va">sz2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">var</a></span><span class="op">(</span><span class="va">y2</span><span class="op">*</span><span class="va">alpha</span><span class="op">)</span> <span class="op">*</span> <span class="op">(</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">h2</span><span class="op">)</span><span class="op">/</span><span class="va">h2</span><span class="op">)</span>
<span class="va">z</span> <span class="op">&lt;-</span> <span class="va">alpha0</span> <span class="op">+</span> <span class="va">y2</span><span class="op">*</span><span class="va">alpha</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">n2</span>,<span class="fl">0</span>,<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="va">sz2</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The genotype data <code>X1</code> and <code>X2</code> are normalized as</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">y</span> <span class="op">=</span> <span class="va">y1</span>;
<span class="va">mean.x1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">X1</span>,<span class="fl">2</span>,<span class="va">mean</span><span class="op">)</span>;
<span class="va">x1m</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html">sweep</a></span><span class="op">(</span><span class="va">X1</span>,<span class="fl">2</span>,<span class="va">mean.x1</span><span class="op">)</span>;
<span class="va">std.x1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">x1m</span>,<span class="fl">2</span>,<span class="va">sd</span><span class="op">)</span>
<span class="va">x1p</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html">sweep</a></span><span class="op">(</span><span class="va">x1m</span>,<span class="fl">2</span>,<span class="va">std.x1</span>,<span class="st">"/"</span><span class="op">)</span>;
<span class="va">x1p</span> <span class="op">=</span> <span class="va">x1p</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">x1p</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>

<span class="va">mean.x2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">X2</span>,<span class="fl">2</span>,<span class="va">mean</span><span class="op">)</span>;
<span class="va">x2m</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html">sweep</a></span><span class="op">(</span><span class="va">X2</span>,<span class="fl">2</span>,<span class="va">mean.x2</span><span class="op">)</span>;
<span class="va">std.x2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">x2m</span>,<span class="fl">2</span>,<span class="va">sd</span><span class="op">)</span>
<span class="va">x2p</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html">sweep</a></span><span class="op">(</span><span class="va">x2m</span>,<span class="fl">2</span>,<span class="va">std.x2</span>,<span class="st">"/"</span><span class="op">)</span>;
<span class="va">x2p</span> <span class="op">=</span> <span class="va">x2p</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">x2p</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>

<span class="va">w2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span>,<span class="va">n2</span><span class="op">)</span>,ncol<span class="op">=</span><span class="fl">1</span><span class="op">)</span>;
<span class="va">w1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span>,<span class="va">n1</span><span class="op">)</span>,ncol<span class="op">=</span><span class="fl">1</span><span class="op">)</span>;</code></pre></div>
<p>Initilize the parameters by using linear mixed model (function <em>lmm_pxem</em>, LMM implemented (n &lt; p) using <a href="https://www.jstor.org/stable/2337481?seq=1#page_scan_tab_contents">PX-EM algorithm</a>, function <em>lmm_pxem2</em>, LMM implemented (n &gt; p)):</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fm0</span> <span class="op">=</span> <span class="fu"><a href="../reference/lmm_pxem2.html">lmm_pxem2</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">w1</span>,<span class="va">x1p</span>, <span class="fl">100</span><span class="op">)</span>
<span class="va">sigma2beta</span> <span class="op">=</span><span class="va">fm0</span><span class="op">$</span><span class="va">sigma2beta</span>;
<span class="va">sigma2y</span> <span class="op">=</span><span class="va">fm0</span><span class="op">$</span><span class="va">sigma2y</span>;
<span class="va">beta0</span> <span class="op">=</span> <span class="va">fm0</span><span class="op">$</span><span class="va">beta0</span>;</code></pre></div>
<p>Fit CoMM w/ and w/o constraint that alpha = 0 as</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fmHa</span> <span class="op">=</span> <span class="fu"><a href="../reference/CoMM_covar_pxem.html">CoMM_covar_pxem</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">z</span>, <span class="va">x1p</span>, <span class="va">x2p</span>, <span class="va">w1</span>, <span class="va">w2</span>,constr <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>;
<span class="va">fmH0</span> <span class="op">=</span> <span class="fu"><a href="../reference/CoMM_covar_pxem.html">CoMM_covar_pxem</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">z</span>, <span class="va">x1p</span>, <span class="va">x2p</span>, <span class="va">w1</span>, <span class="va">w2</span>,constr <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>;
<span class="va">loglikHa</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">fmHa</span><span class="op">$</span><span class="va">loglik</span>,na.rm<span class="op">=</span><span class="cn">T</span><span class="op">)</span>
<span class="va">loglikH0</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">fmH0</span><span class="op">$</span><span class="va">loglik</span>,na.rm<span class="op">=</span><span class="cn">T</span><span class="op">)</span>
<span class="va">tstat</span> <span class="op">=</span> <span class="fl">2</span> <span class="op">*</span> <span class="op">(</span><span class="va">loglikHa</span> <span class="op">-</span> <span class="va">loglikH0</span><span class="op">)</span>;
<span class="va">pval</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Chisquare.html">pchisq</a></span><span class="op">(</span><span class="va">tstat</span>,<span class="fl">1</span>,lower.tail<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="va">alpha_hat</span> <span class="op">=</span> <span class="va">fmHa</span><span class="op">$</span><span class="va">alpha</span> </code></pre></div>
</div>
<div id="fit-comm-using-gwas-and-eqtl-data" class="section level3">
<h3 class="hasAnchor">
<a href="#fit-comm-using-gwas-and-eqtl-data" class="anchor"></a>Fit CoMM using GWAS and eQTL data</h3>
<p>The example of running CoMM using GWAS and eQTL data in plink binary format</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">file1</span> <span class="op">=</span> <span class="st">"1000G.EUR.QC.1"</span>;
<span class="va">file2</span> <span class="op">=</span> <span class="st">"NFBC_filter_mph10"</span>;
<span class="va">file3</span> <span class="op">=</span> <span class="st">"Geuvadis_gene_expression_qn.txt"</span>;
<span class="va">file4</span> <span class="op">=</span> <span class="st">""</span>;
<span class="va">file5</span> <span class="op">=</span> <span class="st">"pc5_NFBC_filter_mph10.txt"</span>;
<span class="va">whichPheno</span> <span class="op">=</span> <span class="fl">1</span>;
<span class="va">bw</span> <span class="op">=</span> <span class="fl">500000</span>;</code></pre></div>
<p>Here, file1 is the prefix for eQTL genotype data in plink binary format, file2 is the GWAS data in plink binary format, file3 is the gene expression file with extended name, file4 and file5 are covariates file for eQTL and GWAS data, respectively. Then run <code>fm = CoMM_testing_run(file1,file2,file3, file4,file5, whichPheno, bw);</code>. For gene expresion file, it must have the following format (rows for genes and columns for individuais and note that it must be tab delimited):</p>
<table class="table">
<colgroup>
<col width="10%">
<col width="10%">
<col width="16%">
<col width="14%">
<col width="20%">
<col width="4%">
<col width="10%">
<col width="11%">
</colgroup>
<thead><tr class="header">
<th align="right">lower</th>
<th align="right">up</th>
<th align="left">genetype1</th>
<th align="left">genetype2</th>
<th align="left">TargetID</th>
<th align="right">Chr</th>
<th align="right">HG00105</th>
<th align="right">HG00115</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">59783540</td>
<td align="right">59843484</td>
<td align="left">lincRNA</td>
<td align="left">PART1</td>
<td align="left">ENSG00000152931.6</td>
<td align="right">5</td>
<td align="right">0.5126086</td>
<td align="right">0.7089508</td>
</tr>
<tr class="even">
<td align="right">48128225</td>
<td align="right">48148330</td>
<td align="left">protein_coding</td>
<td align="left">UPP1</td>
<td align="left">ENSG00000183696.9</td>
<td align="right">7</td>
<td align="right">1.4118007</td>
<td align="right">-0.0135644</td>
</tr>
<tr class="odd">
<td align="right">57846106</td>
<td align="right">57853063</td>
<td align="left">protein_coding</td>
<td align="left">INHBE</td>
<td align="left">ENSG00000139269.2</td>
<td align="right">12</td>
<td align="right">0.5755268</td>
<td align="right">-1.0162217</td>
</tr>
<tr class="even">
<td align="right">116054583</td>
<td align="right">116164515</td>
<td align="left">protein_coding</td>
<td align="left">AFAP1L2</td>
<td align="left">ENSG00000169129.8</td>
<td align="right">10</td>
<td align="right">1.1117776</td>
<td align="right">0.0407033</td>
</tr>
<tr class="odd">
<td align="right">22157909</td>
<td align="right">22396763</td>
<td align="left">protein_coding</td>
<td align="left">RAPGEF5</td>
<td align="left">ENSG00000136237.12</td>
<td align="right">7</td>
<td align="right">0.2831573</td>
<td align="right">-0.1772559</td>
</tr>
<tr class="even">
<td align="right">11700964</td>
<td align="right">11743303</td>
<td align="left">lincRNA</td>
<td align="left">RP11-434C1.1</td>
<td align="left">ENSG00000247157.2</td>
<td align="right">12</td>
<td align="right">0.2550282</td>
<td align="right">-0.2831573</td>
</tr>
</tbody>
</table>
<p>To make ‘CoMM’ further speeding, we implement multiple thread version of ‘CoMM’ by just run <code>fm = CoMM_testing_run_mt(file1,file2,file3, file4,file5, whichPheno, bw, coreNum);</code> where <code>coreNum = 24</code> is the number of cores in your CPU.</p>
</div>
<div id="figures" class="section level3">
<h3 class="hasAnchor">
<a href="#figures" class="anchor"></a>Figures</h3>
<p>The following data and codes are used to produce one of the figures in the Yang et al. (2018).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dat_rej</span> <span class="op">=</span> <span class="va">dat</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>;
<span class="va">dat_rej</span><span class="op">$</span><span class="va">h2z</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">""</span>,<span class="va">dat_rej</span><span class="op">$</span><span class="va">h2</span>,sep<span class="op">=</span><span class="st">""</span><span class="op">)</span>
<span class="va">dat_rej</span><span class="op">$</span><span class="va">Power</span> <span class="op">=</span> <span class="va">dat_rej</span><span class="op">$</span><span class="va">rej_prop</span>
<span class="va">dat_rej</span><span class="op">$</span><span class="va">Sparsity</span> <span class="op">=</span> <span class="va">dat_rej</span><span class="op">$</span><span class="va">beta_prop</span>
<span class="va">dat_rej</span><span class="op">$</span><span class="va">sd_rej</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">dat_rej</span><span class="op">$</span><span class="va">sd_rej</span><span class="op">)</span><span class="op">)</span>
<span class="va">dat_rej</span> <span class="op">=</span> <span class="va">dat_rej</span><span class="op">[</span><span class="va">dat_rej</span><span class="op">$</span><span class="va">Method</span><span class="op">!=</span><span class="st">"2-stage:AUDI"</span>,<span class="op">]</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://had.co.nz/plyr">plyr</a></span><span class="op">)</span>
<span class="va">dat_rej</span><span class="op">$</span><span class="va">Method</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/plyr/man/revalue.html">revalue</a></span><span class="op">(</span><span class="va">dat_rej</span><span class="op">$</span><span class="va">Method</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"AUDI"</span><span class="op">=</span><span class="st">"CoMM"</span><span class="op">)</span><span class="op">)</span>
<span class="va">dat_rej</span><span class="op">$</span><span class="va">Method</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/plyr/man/revalue.html">revalue</a></span><span class="op">(</span><span class="va">dat_rej</span><span class="op">$</span><span class="va">Method</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"2-stage:Ridge"</span><span class="op">=</span><span class="st">"PrediXcan:Ridge"</span><span class="op">)</span><span class="op">)</span>
<span class="va">dat_rej</span><span class="op">$</span><span class="va">Method</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/plyr/man/revalue.html">revalue</a></span><span class="op">(</span><span class="va">dat_rej</span><span class="op">$</span><span class="va">Method</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"2-stage:Enet"</span><span class="op">=</span><span class="st">"PrediXcan:Enet"</span><span class="op">)</span><span class="op">)</span>
<span class="va">dat_rej</span><span class="op">$</span><span class="va">Method</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/droplevels.html">droplevels</a></span><span class="op">(</span><span class="va">dat_rej</span><span class="op">$</span><span class="va">Method</span><span class="op">)</span>

<span class="va">rho</span> <span class="op">=</span> <span class="fl">0.5</span>; <span class="va">n2</span> <span class="op">=</span> <span class="fl">8000</span>;
<span class="va">t1e_rej</span> <span class="op">=</span> <span class="va">dat_rej</span><span class="op">[</span><span class="va">dat_rej</span><span class="op">$</span><span class="va">RhoX</span><span class="op">==</span><span class="va">rho</span><span class="op">&amp;</span><span class="va">dat_rej</span><span class="op">$</span><span class="va">n2</span><span class="op">==</span><span class="va">n2</span>,<span class="op">]</span>

<span class="va">t1e_rej</span><span class="op">$</span><span class="va">h2z</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">t1e_rej</span><span class="op">$</span><span class="va">h2z</span><span class="op">)</span>
<span class="va">t1e_rej</span><span class="op">$</span><span class="va">h2y</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">t1e_rej</span><span class="op">$</span><span class="va">h2y</span><span class="op">)</span>
<span class="va">t1e_rej</span><span class="op">$</span><span class="va">Sparsity</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">t1e_rej</span><span class="op">$</span><span class="va">Sparsity</span><span class="op">)</span>
<span class="va">t1e_rej</span><span class="op">$</span><span class="va">n2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">t1e_rej</span><span class="op">$</span><span class="va">n2</span><span class="op">)</span>
<span class="va">t1e_rej</span><span class="op">$</span><span class="va">Method</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">ordered</a></span><span class="op">(</span><span class="va">t1e_rej</span><span class="op">$</span><span class="va">Method</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CoMM"</span>,<span class="st">"PrediXcan:Ridge"</span>,<span class="st">"PrediXcan:Enet"</span>,<span class="st">"SKAT"</span><span class="op">)</span><span class="op">)</span>
<span class="va">t1e_rej</span><span class="op">$</span><span class="va">Power</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="op">(</span><span class="va">t1e_rej</span><span class="op">$</span><span class="va">Power</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="va">t1e_rej</span><span class="op">$</span><span class="va">h2y2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">t1e_rej</span><span class="op">$</span><span class="va">h2y</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"h[C]^2==0.01"</span>, <span class="st">"h[C]^2==0.03"</span>, 
                      <span class="st">"h[C]^2==0.05"</span>, <span class="st">"h[C]^2==0.07"</span>, <span class="st">"h[C]^2==0.09"</span><span class="op">)</span><span class="op">)</span>
<span class="va">t1e_rej</span><span class="op">$</span><span class="va">h2z2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">t1e_rej</span><span class="op">$</span><span class="va">h2z</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"h[T]^2==0"</span>, <span class="st">"h[T]^2==0.001"</span>, 
                      <span class="st">"h[T]^2==0.002"</span>, <span class="st">"h[T]^2==0.003"</span><span class="op">)</span><span class="op">)</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">t1e_rej</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Sparsity</span>, y <span class="op">=</span> <span class="va">Power</span>,fill <span class="op">=</span> <span class="va">Method</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span>, position<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_dodge.html">position_dodge</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_errorbar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>ymin<span class="op">=</span><span class="va">Power</span><span class="op">-</span><span class="va">sd_rej</span>, ymax<span class="op">=</span><span class="va">Power</span><span class="op">+</span><span class="va">sd_rej</span><span class="op">)</span>, width<span class="op">=</span><span class="fl">.2</span>,
                 position<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_dodge.html">position_dodge</a></span><span class="op">(</span><span class="fl">.9</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span><span class="op">(</span><span class="va">h2z2</span><span class="op">~</span><span class="va">h2y2</span>,labeller <span class="op">=</span> <span class="va">label_parsed</span>,scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept<span class="op">=</span><span class="fl">0.05</span>,colour<span class="op">=</span><span class="st">"orange"</span>,linetype<span class="op">=</span><span class="st">"dashed"</span><span class="op">)</span><span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"bottom"</span><span class="op">)</span></code></pre></div>
<p><img src="CoMM_files/figure-html/unnamed-chunk-11-1.png" width="672"></p>
</div>
<div id="corrections-for-comms-yang-et-al-" class="section level3">
<h3 class="hasAnchor">
<a href="#corrections-for-comms-yang-et-al-" class="anchor"></a>Corrections for CoMMs (Yang et al.)</h3>
<p>In Algorithm 1 (in the supplementary document), the Reduction-step should be <span class="math inline">\(\left( \sigma_u^{(t+1)}\right)^2 = \left( \gamma^{(t+1)}\right)^2\left( \sigma_u^{(t+1)}\right)^2\)</span>.</p>
</div>
<div id="fit-comm_s2-using-simulated-data" class="section level3">
<h3 class="hasAnchor">
<a href="#fit-comm_s2-using-simulated-data" class="anchor"></a>Fit CoMM_S2 using simulated data</h3>
<p>We first generate genotype data using function <em>genRawGeno</em>:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://mvtnorm.R-forge.R-project.org">mvtnorm</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span>
<span class="va">L</span> <span class="op">=</span> <span class="fl">1</span>; <span class="va">M</span> <span class="op">=</span> <span class="fl">100</span>; <span class="va">rho</span> <span class="op">=</span><span class="fl">0.5</span>
<span class="va">n1</span> <span class="op">=</span> <span class="fl">400</span>; <span class="va">n2</span> <span class="op">=</span> <span class="fl">5000</span>; <span class="va">n3</span> <span class="op">=</span> <span class="fl">400</span>;
<span class="va">maf</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="va">M</span>, min <span class="op">=</span> <span class="fl">0.05</span>, max <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>;
<span class="va">X</span> <span class="op">=</span> <span class="fu">genRawGeno</span><span class="op">(</span><span class="va">maf</span>, <span class="va">L</span>, <span class="va">M</span>, <span class="va">rho</span>, <span class="va">n1</span> <span class="op">+</span> <span class="va">n2</span><span class="op">)</span>;
<span class="va">X3</span> <span class="op">=</span> <span class="fu">genRawGeno</span><span class="op">(</span><span class="va">maf</span>, <span class="va">L</span>, <span class="va">M</span>, <span class="va">rho</span>, <span class="va">n3</span><span class="op">)</span></code></pre></div>
<p>Then, effect sizes are generated from standard Gaussian distribution with sparse structure:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">beta_prop</span> <span class="op">=</span> <span class="fl">0.2</span>;
<span class="va">b</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">numeric</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span>;
<span class="va">m</span> <span class="op">=</span> <span class="va">M</span> <span class="op">*</span> <span class="va">beta_prop</span>;
<span class="va">b</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">m</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span>;</code></pre></div>
<p>Subsequently, the gene expression <code>y</code> is generated by controlling cellular heritability at prespecified level (<code>h2y</code>):</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">h2y</span> <span class="op">=</span> <span class="fl">0.05</span>;
<span class="va">b0</span> <span class="op">=</span> <span class="fl">6</span>;
<span class="va">y0</span> <span class="op">&lt;-</span> <span class="va">X</span><span class="op">%*%</span><span class="va">b</span> <span class="op">+</span> <span class="va">b0</span>;
<span class="va">y</span>  <span class="op">&lt;-</span> <span class="va">y0</span> <span class="op">+</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">var</a></span><span class="op">(</span><span class="va">y0</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">h2y</span><span class="op">)</span><span class="op">/</span><span class="va">h2y</span><span class="op">)</span><span class="op">)</span><span class="op">^</span><span class="fl">0.5</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">n1</span><span class="op">+</span><span class="va">n2</span><span class="op">)</span>;</code></pre></div>
<p>Finally, the phenotype data is generated as the generative model of CoMM with a prespecified trait heritability (<code>h2</code>) as:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">h2</span> <span class="op">=</span> <span class="fl">0.001</span>;
<span class="va">y1</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n1</span><span class="op">]</span>
<span class="va">X1</span> <span class="op">&lt;-</span> <span class="va">X</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n1</span>,<span class="op">]</span>
<span class="va">y2</span> <span class="op">&lt;-</span> <span class="va">y0</span><span class="op">[</span><span class="op">(</span><span class="va">n1</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="va">n1</span><span class="op">+</span><span class="va">n2</span><span class="op">)</span><span class="op">]</span>
<span class="va">X2</span> <span class="op">&lt;-</span> <span class="va">X</span><span class="op">[</span><span class="op">(</span><span class="va">n1</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="va">n1</span><span class="op">+</span><span class="va">n2</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">alpha0</span> <span class="op">&lt;-</span> <span class="fl">3</span>  
<span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.3</span>
<span class="va">sz2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">var</a></span><span class="op">(</span><span class="va">y2</span><span class="op">*</span><span class="va">alpha</span><span class="op">)</span> <span class="op">*</span> <span class="op">(</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">h2</span><span class="op">)</span><span class="op">/</span><span class="va">h2</span><span class="op">)</span>
<span class="va">z</span> <span class="op">&lt;-</span> <span class="va">alpha0</span> <span class="op">+</span> <span class="va">y2</span><span class="op">*</span><span class="va">alpha</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">n2</span>,<span class="fl">0</span>,<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="va">sz2</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The genotype data <code>X1</code>, <code>X2</code> and <code>X3</code> are centered as</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">y</span> <span class="op">=</span> <span class="va">y1</span>;
<span class="va">mean.x1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">X1</span>,<span class="fl">2</span>,<span class="va">mean</span><span class="op">)</span>;
<span class="va">x1p</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html">sweep</a></span><span class="op">(</span><span class="va">X1</span>,<span class="fl">2</span>,<span class="va">mean.x1</span><span class="op">)</span>;

<span class="va">mean.x2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">X2</span>,<span class="fl">2</span>,<span class="va">mean</span><span class="op">)</span>;
<span class="va">x2p</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html">sweep</a></span><span class="op">(</span><span class="va">X2</span>,<span class="fl">2</span>,<span class="va">mean.x2</span><span class="op">)</span>;

<span class="va">mean.x3</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">X3</span>,<span class="fl">2</span>,<span class="va">mean</span><span class="op">)</span>;
<span class="va">x3p</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html">sweep</a></span><span class="op">(</span><span class="va">X3</span>,<span class="fl">2</span>,<span class="va">mean.x3</span><span class="op">)</span>;

<span class="va">w</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span>,<span class="va">n1</span><span class="op">)</span>,ncol<span class="op">=</span><span class="fl">1</span><span class="op">)</span>;</code></pre></div>
<p>The summary statistics are generated from GWAS individual data</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hatmu</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">M</span>, <span class="fl">1</span><span class="op">)</span>
<span class="va">hats</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">M</span>, <span class="fl">1</span><span class="op">)</span>

<span class="kw">for</span> <span class="op">(</span><span class="va">m</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">M</span><span class="op">)</span><span class="op">{</span>
  <span class="va">fm</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">z</span><span class="op">~</span><span class="fl">1</span><span class="op">+</span><span class="va">x2p</span><span class="op">[</span>,<span class="va">m</span><span class="op">]</span><span class="op">)</span>;
  <span class="va">hatmu</span><span class="op">[</span><span class="va">m</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">fm</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">]</span>
  <span class="va">hats</span><span class="op">[</span><span class="va">m</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">fm</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">]</span>;
<span class="op">}</span></code></pre></div>
<p>The correlation matrix reflecting LD information is estimated using reference panel</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lam</span> <span class="op">=</span> <span class="fl">0.8</span>
<span class="va">sumx3p</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">x3p</span><span class="op">*</span><span class="va">x3p</span>, <span class="fl">2</span>, <span class="va">sum</span><span class="op">)</span>
<span class="va">R</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">M</span>, <span class="va">M</span><span class="op">)</span>;
<span class="kw">for</span> <span class="op">(</span><span class="va">i1</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">M</span><span class="op">)</span><span class="op">{</span>
  <span class="kw">for</span> <span class="op">(</span><span class="va">j1</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">M</span><span class="op">)</span><span class="op">{</span>
    <span class="va">R</span><span class="op">[</span><span class="va">i1</span>,<span class="va">j1</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">x3p</span><span class="op">[</span>,<span class="va">i1</span><span class="op">]</span><span class="op">)</span><span class="op">%*%</span><span class="va">x3p</span><span class="op">[</span>,<span class="va">j1</span><span class="op">]</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="va">sumx3p</span><span class="op">[</span><span class="va">i1</span><span class="op">]</span><span class="op">*</span><span class="va">sumx3p</span><span class="op">[</span><span class="va">j1</span><span class="op">]</span><span class="op">)</span>                        
  <span class="op">}</span>
<span class="op">}</span>
<span class="va">R</span> <span class="op">=</span> <span class="va">R</span><span class="op">*</span><span class="va">lam</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">lam</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span>  </code></pre></div>
<p>The likelihood ratio test is implemented</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">px</span> <span class="op">=</span> <span class="fl">1</span>
<span class="va">opts</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>max_iter <span class="op">=</span> <span class="fl">10000</span>, dispF <span class="op">=</span> <span class="fl">1</span>, display_gap <span class="op">=</span> <span class="fl">10</span>, epsStopLogLik <span class="op">=</span> <span class="fl">1e-5</span>, fix_alphag <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>;
<span class="va">opts1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>max_iter <span class="op">=</span> <span class="fl">10000</span>, dispF <span class="op">=</span> <span class="fl">1</span>, display_gap <span class="op">=</span> <span class="fl">10</span>, epsStopLogLik <span class="op">=</span> <span class="fl">1e-5</span>, fix_alphag <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>;

<span class="va">fmHa</span> <span class="op">=</span> <span class="fu"><a href="../reference/CoMM_S2.html">CoMM_S2</a></span><span class="op">(</span><span class="va">x1p</span>, <span class="va">y</span>, <span class="va">w</span>, <span class="va">hatmu</span>, <span class="va">hats</span>, <span class="va">R</span>, <span class="va">opts</span>, <span class="va">px</span><span class="op">)</span>;
<span class="co">#&gt; ***Iteration*******Fnew********Fold**********Diff***</span>
<span class="co">#&gt;    1.0000e+01  -1.1037e+03  -1.1037e+03   1.4388e-02</span>
<span class="co">#&gt; ***Iteration*******Fnew********Fold**********Diff***</span>
<span class="co">#&gt;    2.0000e+01  -1.1036e+03  -1.1036e+03   3.4670e-05</span>
<span class="va">fmH0</span> <span class="op">=</span> <span class="fu"><a href="../reference/CoMM_S2.html">CoMM_S2</a></span><span class="op">(</span><span class="va">x1p</span>, <span class="va">y</span>, <span class="va">w</span>, <span class="va">hatmu</span>, <span class="va">hats</span>, <span class="va">R</span>, <span class="va">opts1</span>, <span class="va">px</span><span class="op">)</span>;
<span class="co">#&gt; ***Iteration*******Fnew********Fold**********Diff***</span>
<span class="co">#&gt;    1.0000e+01  -1.1048e+03  -1.1048e+03   2.9406e-03</span>

<span class="va">stat</span> <span class="op">=</span> <span class="fl">2</span><span class="op">*</span><span class="op">(</span><span class="va">fmHa</span><span class="op">$</span><span class="va">LRLB</span> <span class="op">-</span> <span class="va">fmH0</span><span class="op">$</span><span class="va">LRLB</span><span class="op">)</span>
<span class="va">pval</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Chisquare.html">pchisq</a></span><span class="op">(</span><span class="va">stat</span>, <span class="fl">1</span>, lower.tail <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">fmHa</span><span class="op">)</span>
<span class="co">#&gt; List of 7</span>
<span class="co">#&gt;  $ vardist_mu: num [1:100, 1] -0.0127 -0.0902 0.015 0.0462 0.0315 ...</span>
<span class="co">#&gt;  $ sigma2mu  : num 0.0459</span>
<span class="co">#&gt;  $ alphag    : num 0.736</span>
<span class="co">#&gt;  $ sigma2beta: num 0.0658</span>
<span class="co">#&gt;  $ sigma2y   : num 69.6</span>
<span class="co">#&gt;  $ LRLB      : num -1161</span>
<span class="co">#&gt;  $ Lq        : num [1, 1:23] -1273 -1109 -1105 -1104 -1104 ...</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">fmH0</span><span class="op">)</span>
<span class="co">#&gt; List of 7</span>
<span class="co">#&gt;  $ vardist_mu: num [1:100, 1] -0.02838 0.00194 -0.03273 -0.07739 -0.05696 ...</span>
<span class="co">#&gt;  $ sigma2mu  : num 0.0469</span>
<span class="co">#&gt;  $ alphag    : num 0</span>
<span class="co">#&gt;  $ sigma2beta: num 0.0658</span>
<span class="co">#&gt;  $ sigma2y   : num 69.6</span>
<span class="co">#&gt;  $ LRLB      : num -1162</span>
<span class="co">#&gt;  $ Lq        : num [1, 1:19] -1274 -1110 -1106 -1105 -1105 ...</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">stat</span><span class="op">)</span>
<span class="co">#&gt; [1] 2.532153</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">pval</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.1115479</span></code></pre></div>
<p>The output of CoMM_S2 is a list with 7 variables, mean of variational distribution <code>vardist_mu</code>, variance component <code>sigma2mu</code>, gene effect size <code>alphag</code>, variance component <code>sigma2y</code>, calibrated ELBO <code>LRLB</code>, original ELBO <code>Lq</code>.</p>
</div>
<div id="fit-comm_s2-using-gwas-and-eqtl-data" class="section level3">
<h3 class="hasAnchor">
<a href="#fit-comm_s2-using-gwas-and-eqtl-data" class="anchor"></a>Fit CoMM_S2 using GWAS and eQTL data</h3>
<p>The example of running CoMM_S2 using GWAS summary statistics and eQTL data in plink binary format</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">file1</span> <span class="op">=</span> <span class="st">"1000G.EUR.QC.1"</span>;
<span class="va">file2</span> <span class="op">=</span> <span class="st">"NFBC_beta_se_TG.txt"</span>
<span class="va">file3</span> <span class="op">=</span> <span class="st">"1000G_chr_all"</span>;
<span class="va">file4</span> <span class="op">=</span> <span class="st">"Geuvadis_gene_expression_qn.txt"</span>;
<span class="va">file5</span> <span class="op">=</span> <span class="st">""</span>;
<span class="va">bw</span> <span class="op">=</span> <span class="fl">500000</span>;
<span class="va">lam</span> <span class="op">=</span> <span class="fl">0.95</span>;
<span class="va">coreNum</span> <span class="op">=</span> <span class="fl">24</span>;</code></pre></div>
<p>Here, file1 is the prefix for eQTL genotype data in plink binary format, file2 is the GWAS summary data, file3 is the prefix for reference panel data in plink binary format, file4 is the gene expression file with extended name, file5 are covariates file for eQTL data. bw is the number of downstream and upstream SNPs that are considered as cis-SNP within a gene. lam is the shirnkage intensify for reference panel. coreNum is the number of cores in parallel. Then run <code>fm = CoMM_S2_testing(file1, file2, file3, file4, file5, bw, lam);</code>. For GWAS summary data file, it must have the following format (note that it must be tab delimited):</p>
<table class="table">
<thead><tr class="header">
<th align="left">SNP</th>
<th align="right">chr</th>
<th align="right">BP</th>
<th align="left">A1</th>
<th align="left">A2</th>
<th align="right">beta</th>
<th align="right">se</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">rs3094315</td>
<td align="right">1</td>
<td align="right">752566</td>
<td align="left">G</td>
<td align="left">A</td>
<td align="right">-0.0122</td>
<td align="right">0.0294</td>
</tr>
<tr class="even">
<td align="left">rs3128117</td>
<td align="right">1</td>
<td align="right">944564</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">-0.0208</td>
<td align="right">0.0278</td>
</tr>
<tr class="odd">
<td align="left">rs1891906</td>
<td align="right">1</td>
<td align="right">950243</td>
<td align="left">C</td>
<td align="left">A</td>
<td align="right">-0.0264</td>
<td align="right">0.0260</td>
</tr>
<tr class="even">
<td align="left">rs2710888</td>
<td align="right">1</td>
<td align="right">959842</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">-0.0439</td>
<td align="right">0.0297</td>
</tr>
<tr class="odd">
<td align="left">rs4970393</td>
<td align="right">1</td>
<td align="right">962606</td>
<td align="left">G</td>
<td align="left">A</td>
<td align="right">-0.0252</td>
<td align="right">0.0233</td>
</tr>
<tr class="even">
<td align="left">rs7526076</td>
<td align="right">1</td>
<td align="right">998395</td>
<td align="left">A</td>
<td align="left">G</td>
<td align="right">-0.0512</td>
<td align="right">0.0229</td>
</tr>
<tr class="odd">
<td align="left">rs4075116</td>
<td align="right">1</td>
<td align="right">1003629</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">-0.0497</td>
<td align="right">0.0220</td>
</tr>
<tr class="even">
<td align="left">rs3934834</td>
<td align="right">1</td>
<td align="right">1005806</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">0.0364</td>
<td align="right">0.0256</td>
</tr>
<tr class="odd">
<td align="left">rs3766192</td>
<td align="right">1</td>
<td align="right">1017197</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">-0.0116</td>
<td align="right">0.0178</td>
</tr>
<tr class="even">
<td align="left">rs3766191</td>
<td align="right">1</td>
<td align="right">1017587</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">0.0318</td>
<td align="right">0.0262</td>
</tr>
</tbody>
</table>
<p>To make ‘CoMM_S2’ further speeding, we implement multiple thread version of ‘CoMM_S2’ by just run <code>fm = CoMM_S2_paral_testing(file1, file2, file3, file4, file5, bw, lam, coreNum);</code></p>
</div>
<div id="figures-1" class="section level3">
<h3 class="hasAnchor">
<a href="#figures-1" class="anchor"></a>Figures</h3>
<p>The following data and codes are used to produce the barplot of power</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://colorspace.R-Forge.R-project.org">colorspace</a></span><span class="op">)</span>
<span class="va">bp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">pval2</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">Sparsity</span>, y<span class="op">=</span><span class="va">Power</span>, fill<span class="op">=</span><span class="va">Method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span>, position<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_dodge.html">position_dodge</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span><span class="op">(</span><span class="va">h2</span><span class="op">~</span><span class="va">hc</span>, scales <span class="op">=</span> <span class="st">"free"</span>, labeller <span class="op">=</span> <span class="va">label_parsed</span><span class="op">)</span>  <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>strip.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span>, color<span class="op">=</span><span class="st">"black"</span>,
                                      face<span class="op">=</span><span class="st">"bold"</span><span class="op">)</span>,
          strip.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span>, color<span class="op">=</span><span class="st">"black"</span>,
                                      face<span class="op">=</span><span class="st">"bold"</span><span class="op">)</span>,
          plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">20</span>,face <span class="op">=</span> <span class="st">"bold"</span>,hjust<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span>,
          axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">8</span>,face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>,
          axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">8</span>,face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>,
          axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
          axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">15</span>,face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>,
          legend.position<span class="op">=</span><span class="st">"bottom"</span>,
          legend.title<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">15</span><span class="op">)</span>,
          legend.text<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span>
  <span class="va">colours</span><span class="op">&lt;-</span><span class="fu"><a href="https://colorspace.R-Forge.R-project.org//reference/rainbow_hcl.html">rainbow_hcl</a></span><span class="op">(</span><span class="fl">3</span>, start <span class="op">=</span> <span class="fl">0</span>, end <span class="op">=</span> <span class="fl">300</span><span class="op">)</span>
  <span class="va">bp2</span> <span class="op">=</span> <span class="va">bp2</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="va">colours</span>, labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/expression.html">expression</a></span><span class="op">(</span><span class="st">"CoMM-S"</span><span class="op">^</span><span class="fl">2</span>,<span class="st">"S-PrediXcan:Ridge"</span>,<span class="st">"S-PrediXcan:Enet"</span><span class="op">)</span><span class="op">)</span>
<span class="va">bp2</span></code></pre></div>
<p><img src="CoMM_files/figure-html/unnamed-chunk-22-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div id="fit-comm_s4-using-simulated-data" class="section level3">
<h3 class="hasAnchor">
<a href="#fit-comm_s4-using-simulated-data" class="anchor"></a>Fit CoMM_S4 using simulated data</h3>
<p>We first generate genotype data using function <em>genRawGeno</em>:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://mvtnorm.R-forge.R-project.org">mvtnorm</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span>
<span class="va">L</span> <span class="op">=</span> <span class="fl">1</span>; <span class="va">M</span> <span class="op">=</span> <span class="fl">100</span>; <span class="va">rho</span> <span class="op">=</span><span class="fl">0.5</span>
<span class="va">n1</span> <span class="op">=</span> <span class="fl">5000</span>; <span class="va">n2</span> <span class="op">=</span> <span class="fl">5000</span>; <span class="va">n3</span> <span class="op">=</span> <span class="fl">400</span>; <span class="va">n4</span> <span class="op">=</span> <span class="fl">400</span>;
<span class="va">maf</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="va">M</span>, min <span class="op">=</span> <span class="fl">0.05</span>, max <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>;
<span class="va">X</span> <span class="op">=</span> <span class="fu">genRawGeno</span><span class="op">(</span><span class="va">maf</span>, <span class="va">L</span>, <span class="va">M</span>, <span class="va">rho</span>, <span class="va">n1</span> <span class="op">+</span> <span class="va">n2</span><span class="op">)</span>;
<span class="va">X3</span> <span class="op">=</span> <span class="fu">genRawGeno</span><span class="op">(</span><span class="va">maf</span>, <span class="va">L</span>, <span class="va">M</span>, <span class="va">rho</span>, <span class="va">n3</span><span class="op">)</span>;
<span class="va">X4</span> <span class="op">=</span> <span class="fu">genRawGeno</span><span class="op">(</span><span class="va">maf</span>, <span class="va">L</span>, <span class="va">M</span>, <span class="va">rho</span>, <span class="va">n4</span><span class="op">)</span></code></pre></div>
<p>Then, the vector of effect size is generated from standard Gaussian distribution with sparse structure:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">beta_prop</span> <span class="op">=</span> <span class="fl">0.2</span>;
<span class="va">b</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">numeric</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span>;
<span class="va">m</span> <span class="op">=</span> <span class="va">M</span> <span class="op">*</span> <span class="va">beta_prop</span>;
<span class="va">b</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">m</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span>;</code></pre></div>
<p>Subsequently, the gene expression <code>y</code> is generated by controlling cellular heritability at prespecified level (<code>h2y</code>):</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">h2y</span> <span class="op">=</span> <span class="fl">0.05</span>;
<span class="va">b0</span> <span class="op">=</span> <span class="fl">6</span>;
<span class="va">y0</span> <span class="op">&lt;-</span> <span class="va">X</span><span class="op">%*%</span><span class="va">b</span> <span class="op">+</span> <span class="va">b0</span>;
<span class="va">y</span>  <span class="op">&lt;-</span> <span class="va">y0</span> <span class="op">+</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">var</a></span><span class="op">(</span><span class="va">y0</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">h2y</span><span class="op">)</span><span class="op">/</span><span class="va">h2y</span><span class="op">)</span><span class="op">)</span><span class="op">^</span><span class="fl">0.5</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">n1</span><span class="op">+</span><span class="va">n2</span><span class="op">)</span>;</code></pre></div>
<p>Finally, the phenotype data is generated as the generative model of CoMM with a prespecified trait heritability (<code>h2</code>) as:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">h2</span> <span class="op">=</span> <span class="fl">0.001</span>;
<span class="va">y1</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n1</span><span class="op">]</span>
<span class="va">X1</span> <span class="op">&lt;-</span> <span class="va">X</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n1</span>,<span class="op">]</span>
<span class="va">y2</span> <span class="op">&lt;-</span> <span class="va">y0</span><span class="op">[</span><span class="op">(</span><span class="va">n1</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="va">n1</span><span class="op">+</span><span class="va">n2</span><span class="op">)</span><span class="op">]</span>
<span class="va">X2</span> <span class="op">&lt;-</span> <span class="va">X</span><span class="op">[</span><span class="op">(</span><span class="va">n1</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="va">n1</span><span class="op">+</span><span class="va">n2</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">alpha0</span> <span class="op">&lt;-</span> <span class="fl">3</span>  
<span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.3</span>
<span class="va">sz2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">var</a></span><span class="op">(</span><span class="va">y2</span><span class="op">*</span><span class="va">alpha</span><span class="op">)</span> <span class="op">*</span> <span class="op">(</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">h2</span><span class="op">)</span><span class="op">/</span><span class="va">h2</span><span class="op">)</span>
<span class="va">z</span> <span class="op">&lt;-</span> <span class="va">alpha0</span> <span class="op">+</span> <span class="va">y2</span><span class="op">*</span><span class="va">alpha</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">n2</span>,<span class="fl">0</span>,<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="va">sz2</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The genotype data <code>X1</code>, <code>X2</code> and <code>X3</code> are centered as</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">y</span> <span class="op">=</span> <span class="va">y1</span>;
<span class="va">mean.x1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">X1</span>,<span class="fl">2</span>,<span class="va">mean</span><span class="op">)</span>;
<span class="va">x1p</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html">sweep</a></span><span class="op">(</span><span class="va">X1</span>,<span class="fl">2</span>,<span class="va">mean.x1</span><span class="op">)</span>;

<span class="va">mean.x2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">X2</span>,<span class="fl">2</span>,<span class="va">mean</span><span class="op">)</span>;
<span class="va">x2p</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html">sweep</a></span><span class="op">(</span><span class="va">X2</span>,<span class="fl">2</span>,<span class="va">mean.x2</span><span class="op">)</span>;

<span class="va">mean.x3</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">X3</span>,<span class="fl">2</span>,<span class="va">mean</span><span class="op">)</span>;
<span class="va">x3p</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html">sweep</a></span><span class="op">(</span><span class="va">X3</span>,<span class="fl">2</span>,<span class="va">mean.x3</span><span class="op">)</span>;

<span class="va">mean.x4</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">X4</span>,<span class="fl">2</span>,<span class="va">mean</span><span class="op">)</span>;
<span class="va">x4p</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html">sweep</a></span><span class="op">(</span><span class="va">X4</span>,<span class="fl">2</span>,<span class="va">mean.x4</span><span class="op">)</span>;</code></pre></div>
<p>The summary statistics are generated from eQTL and GWAS individual data</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hatmu</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">M</span>, <span class="fl">1</span><span class="op">)</span>
<span class="va">hats</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">M</span>, <span class="fl">1</span><span class="op">)</span>

<span class="kw">for</span> <span class="op">(</span><span class="va">m</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">M</span><span class="op">)</span><span class="op">{</span>
  <span class="va">fm</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">y</span><span class="op">~</span><span class="fl">1</span><span class="op">+</span><span class="va">x2p</span><span class="op">[</span>,<span class="va">m</span><span class="op">]</span><span class="op">)</span>;
  <span class="va">hatmu</span><span class="op">[</span><span class="va">m</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">fm</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">]</span>;
  <span class="va">hats</span><span class="op">[</span><span class="va">m</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">fm</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">]</span>;
<span class="op">}</span>

<span class="va">hatmu2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">M</span>, <span class="fl">1</span><span class="op">)</span>
<span class="va">hats2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">M</span>, <span class="fl">1</span><span class="op">)</span>

<span class="kw">for</span> <span class="op">(</span><span class="va">m</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">M</span><span class="op">)</span><span class="op">{</span>
  <span class="va">fm</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">z</span><span class="op">~</span><span class="fl">1</span><span class="op">+</span><span class="va">x2p</span><span class="op">[</span>,<span class="va">m</span><span class="op">]</span><span class="op">)</span>;
  <span class="va">hatmu2</span><span class="op">[</span><span class="va">m</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">fm</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">]</span>;
  <span class="va">hats2</span><span class="op">[</span><span class="va">m</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">fm</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">]</span>;
<span class="op">}</span></code></pre></div>
<p>The correlation matrix reflecting LD information is estimated using reference panel</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lam</span> <span class="op">=</span> <span class="fl">0.8</span>
<span class="va">sumx3p</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">x3p</span><span class="op">*</span><span class="va">x3p</span>, <span class="fl">2</span>, <span class="va">sum</span><span class="op">)</span>
<span class="va">R</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">M</span>, <span class="va">M</span><span class="op">)</span>;
<span class="kw">for</span> <span class="op">(</span><span class="va">i1</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">M</span><span class="op">)</span><span class="op">{</span>
  <span class="kw">for</span> <span class="op">(</span><span class="va">j1</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">M</span><span class="op">)</span><span class="op">{</span>
    <span class="va">R</span><span class="op">[</span><span class="va">i1</span>,<span class="va">j1</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">x3p</span><span class="op">[</span>,<span class="va">i1</span><span class="op">]</span><span class="op">)</span><span class="op">%*%</span><span class="va">x3p</span><span class="op">[</span>,<span class="va">j1</span><span class="op">]</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="va">sumx3p</span><span class="op">[</span><span class="va">i1</span><span class="op">]</span><span class="op">*</span><span class="va">sumx3p</span><span class="op">[</span><span class="va">j1</span><span class="op">]</span><span class="op">)</span>                        
  <span class="op">}</span>
<span class="op">}</span>
<span class="va">R</span> <span class="op">=</span> <span class="va">R</span><span class="op">*</span><span class="va">lam</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">lam</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span> 


<span class="va">sumx4p</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">x4p</span><span class="op">*</span><span class="va">x4p</span>, <span class="fl">2</span>, <span class="va">sum</span><span class="op">)</span>
<span class="va">R2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">M</span>, <span class="va">M</span><span class="op">)</span>;
<span class="kw">for</span> <span class="op">(</span><span class="va">i1</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">M</span><span class="op">)</span><span class="op">{</span>
  <span class="kw">for</span> <span class="op">(</span><span class="va">j1</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">M</span><span class="op">)</span><span class="op">{</span>
    <span class="va">R2</span><span class="op">[</span><span class="va">i1</span>,<span class="va">j1</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">x4p</span><span class="op">[</span>,<span class="va">i1</span><span class="op">]</span><span class="op">)</span><span class="op">%*%</span><span class="va">x4p</span><span class="op">[</span>,<span class="va">j1</span><span class="op">]</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="va">sumx4p</span><span class="op">[</span><span class="va">i1</span><span class="op">]</span><span class="op">*</span><span class="va">sumx4p</span><span class="op">[</span><span class="va">j1</span><span class="op">]</span><span class="op">)</span>                        
  <span class="op">}</span>
<span class="op">}</span>
<span class="va">R2</span> <span class="op">=</span> <span class="va">R2</span><span class="op">*</span><span class="va">lam</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">lam</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span> </code></pre></div>
<p>The likelihood ratio test is implemented</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">px</span> <span class="op">=</span> <span class="fl">1</span>
<span class="va">opts</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>max_iter <span class="op">=</span> <span class="fl">10000</span>, dispF <span class="op">=</span> <span class="fl">1</span>, display_gap <span class="op">=</span> <span class="fl">10</span>, epsStopLogLik <span class="op">=</span> <span class="fl">1e-5</span>, fix_alphag <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>;
<span class="va">opts1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>max_iter <span class="op">=</span> <span class="fl">10000</span>, dispF <span class="op">=</span> <span class="fl">1</span>, display_gap <span class="op">=</span> <span class="fl">10</span>, epsStopLogLik <span class="op">=</span> <span class="fl">1e-5</span>, fix_alphag <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>;

<span class="va">zscore</span> <span class="op">=</span> <span class="va">hatmu</span><span class="op">/</span><span class="va">hats</span>
<span class="va">zscore2</span> <span class="op">=</span> <span class="va">hatmu2</span><span class="op">/</span><span class="va">hats2</span>

<span class="va">fmHa</span> <span class="op">=</span> <span class="fu"><a href="../reference/CoMM_S4.html">CoMM_S4</a></span><span class="op">(</span><span class="va">zscore</span>, <span class="va">zscore2</span>, <span class="va">R</span>, <span class="va">R2</span>, <span class="va">opts</span>, <span class="va">px</span><span class="op">)</span>;
<span class="co">#&gt; ***Iteration*******Fnew********Fold**********Diff***</span>
<span class="co">#&gt;    10.0000  -49.9045  -49.9073    0.0027</span>
<span class="co">#&gt; ***Iteration*******Fnew********Fold**********Diff***</span>
<span class="co">#&gt;    2.0000e+01  -4.9900e+01  -4.9900e+01   4.0855e-05</span>
<span class="va">fmH0</span> <span class="op">=</span> <span class="fu"><a href="../reference/CoMM_S4.html">CoMM_S4</a></span><span class="op">(</span><span class="va">zscore</span>, <span class="va">zscore2</span>, <span class="va">R</span>, <span class="va">R2</span>, <span class="va">opts1</span>, <span class="va">px</span><span class="op">)</span>;
<span class="co">#&gt; ***Iteration*******Fnew********Fold**********Diff***</span>
<span class="co">#&gt;    10.0000  -50.0588  -50.0732    0.0144</span>
<span class="co">#&gt; ***Iteration*******Fnew********Fold**********Diff***</span>
<span class="co">#&gt;    20.0000  -50.0112  -50.0130    0.0017</span>
<span class="co">#&gt; ***Iteration*******Fnew********Fold**********Diff***</span>
<span class="co">#&gt;    30.0000  -50.0030  -50.0034    0.0004</span>
<span class="co">#&gt; ***Iteration*******Fnew********Fold**********Diff***</span>
<span class="co">#&gt;    40.0000  -50.0009  -50.0010    0.0001</span>
<span class="co">#&gt; ***Iteration*******Fnew********Fold**********Diff***</span>
<span class="co">#&gt;    5.0000e+01  -5.0000e+01  -5.0000e+01   3.2494e-05</span>
<span class="co">#&gt; ***Iteration*******Fnew********Fold**********Diff***</span>
<span class="co">#&gt;    6.0000e+01  -5.0000e+01  -5.0000e+01   9.5248e-06</span>

<span class="va">stat</span> <span class="op">=</span> <span class="fl">2</span><span class="op">*</span><span class="op">(</span><span class="va">fmHa</span><span class="op">$</span><span class="va">LRLB</span> <span class="op">-</span> <span class="va">fmH0</span><span class="op">$</span><span class="va">LRLB</span><span class="op">)</span>
<span class="va">pval</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Chisquare.html">pchisq</a></span><span class="op">(</span><span class="va">stat</span>, <span class="fl">1</span>, lower.tail <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">fmHa</span><span class="op">)</span>
<span class="co">#&gt; List of 5</span>
<span class="co">#&gt;  $ vardist_mu: num [1:100, 1] -0.00298 0.01132 0.11214 0.07152 -0.02412 ...</span>
<span class="co">#&gt;  $ sigma2mu  : num [1, 1] 0.0279</span>
<span class="co">#&gt;  $ alphag    : num -0.936</span>
<span class="co">#&gt;  $ Lq        : num [1, 1:25] -54.3 -51 -50.3 -50.1 -50 ...</span>
<span class="co">#&gt;  $ LRLB      : num -49.9</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">fmH0</span><span class="op">)</span>
<span class="co">#&gt; List of 5</span>
<span class="co">#&gt;  $ vardist_mu: num [1:100, 1] -1.02e-05 6.16e-06 3.74e-05 1.80e-05 -2.40e-05 ...</span>
<span class="co">#&gt;  $ sigma2mu  : num [1, 1] 2.32e-05</span>
<span class="co">#&gt;  $ alphag    : num 0</span>
<span class="co">#&gt;  $ Lq        : num [1, 1:60] -54.8 -51.4 -50.6 -50.4 -50.2 ...</span>
<span class="co">#&gt;  $ LRLB      : num -50</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">stat</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.2349252</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">pval</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.6278957</span></code></pre></div>
<p>The output of CoMM_S4 is a list with 5 variables, mean of variational distribution <code>vardist_mu</code>, variance component <code>sigma2mu</code>, gene effect size <code>alphag</code>, calibrated ELBO <code>LRLB</code>, original ELBO <code>Lq</code>.</p>
</div>
<div id="fit-comm_s4-using-eqtl-and-gwas-summary-statistics" class="section level3">
<h3 class="hasAnchor">
<a href="#fit-comm_s4-using-eqtl-and-gwas-summary-statistics" class="anchor"></a>Fit CoMM_S4 using eQTL and GWAS summary statistics</h3>
<p>The example of running TransCoMM_S2 using eQTL summary statistics and GWAS summary statistics</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">stringname1</span> <span class="op">=</span> <span class="st">"cis-eQTLs_full_chr1_1-100000.txt"</span>;
<span class="va">stringname2</span> <span class="op">=</span> <span class="st">"NFBC_ph1_beta_se.txt"</span>
<span class="va">stringname3</span> <span class="op">=</span> <span class="st">"all_chr_1000G"</span>;
<span class="va">stringname4</span> <span class="op">=</span> <span class="st">"cis-eQTLs_full_chr1"</span>;
<span class="va">stringname5</span> <span class="op">=</span> <span class="st">"all_chr_1000G"</span>;
<span class="va">px</span> <span class="op">=</span> <span class="fl">1</span>
<span class="va">lam</span> <span class="op">=</span> <span class="fl">0.95</span>;
<span class="va">coreNum</span> <span class="op">=</span> <span class="fl">24</span>;</code></pre></div>
<p>Here, the parameter stringname1 is the eQTL summary statistics. The parameter stringname2 is the GWAS summary statistics whose format is the same as the GWAS summary statistics of the input file for model CoMM_S2. The parameter stringname3 is the prefix for reference panel of eQTL in plink binary format. The parameter stringname4 is the prefix for the file to complement additional information of SNPs in eQTL summary statistics file, it has the same format as bim file used in plink software. The parameter stringname5 is the prefix for the reference panel of GWAS in plink binary format. The parameter px is a logical value (1 or 0) indicating whether to use paramter expansion to accerate algorithm. The parameter lam is the shrinkage intensity for the reference panel. The parameter coreNum is the number of cores in parallel. Then run <code>fm = CoMM_S4_testing(stringname1, stringname2, stringname3, stringname4, stringname5, px, lam);</code>.</p>
<p>For the eQTL summary data file, it requires to have the following two formats (note that it must be tab delimited):</p>
<table class="table">
<thead><tr class="header">
<th align="left">SNP</th>
<th align="left">Gene</th>
<th align="right">beta</th>
<th align="right">se</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">rs201725126</td>
<td align="left">ENSG00000227232</td>
<td align="right">0.6142857</td>
<td align="right">0.6780778</td>
</tr>
<tr class="even">
<td align="left">rs200579949</td>
<td align="left">ENSG00000227232</td>
<td align="right">-0.7585898</td>
<td align="right">0.0453431</td>
</tr>
<tr class="odd">
<td align="left">rs75454623</td>
<td align="left">ENSG00000227232</td>
<td align="right">0.2414931</td>
<td align="right">0.4935305</td>
</tr>
<tr class="even">
<td align="left">rs199856693</td>
<td align="left">ENSG00000170113</td>
<td align="right">0.5050623</td>
<td align="right">0.2252596</td>
</tr>
<tr class="odd">
<td align="left">rs78601809</td>
<td align="left">ENSG00000170113</td>
<td align="right">1.5409776</td>
<td align="right">0.7169221</td>
</tr>
<tr class="even">
<td align="left">rs200482301</td>
<td align="left">ENSG00000170113</td>
<td align="right">0.2524854</td>
<td align="right">0.2358414</td>
</tr>
<tr class="odd">
<td align="left">rs140337953</td>
<td align="left">ENSG00000140157</td>
<td align="right">0.5835819</td>
<td align="right">0.2434627</td>
</tr>
<tr class="even">
<td align="left">rs200943160</td>
<td align="left">ENSG00000140157</td>
<td align="right">0.0126301</td>
<td align="right">0.6551085</td>
</tr>
<tr class="odd">
<td align="left">rs116400033</td>
<td align="left">ENSG00000140157</td>
<td align="right">0.8584214</td>
<td align="right">1.7336470</td>
</tr>
<tr class="even">
<td align="left">rs61999471</td>
<td align="left">ENSG00000140157</td>
<td align="right">-0.3483619</td>
<td align="right">1.8034719</td>
</tr>
</tbody>
</table>
<p>or</p>
<table class="table">
<thead><tr class="header">
<th align="left">SNP</th>
<th align="left">Gene</th>
<th align="right">zscore</th>
<th align="right">constant</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">rs201725126</td>
<td align="left">ENSG00000227232</td>
<td align="right">-0.3778055</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">rs200579949</td>
<td align="left">ENSG00000227232</td>
<td align="right">-0.2706801</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">rs75454623</td>
<td align="left">ENSG00000227232</td>
<td align="right">0.7614850</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">rs199856693</td>
<td align="left">ENSG00000170113</td>
<td align="right">-0.0505628</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">rs78601809</td>
<td align="left">ENSG00000170113</td>
<td align="right">1.3861256</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">rs200482301</td>
<td align="left">ENSG00000170113</td>
<td align="right">-1.2418526</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">rs140337953</td>
<td align="left">ENSG00000140157</td>
<td align="right">0.6853001</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">rs200943160</td>
<td align="left">ENSG00000140157</td>
<td align="right">-0.3329476</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">rs116400033</td>
<td align="left">ENSG00000140157</td>
<td align="right">1.1296831</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">rs61999471</td>
<td align="left">ENSG00000140157</td>
<td align="right">1.3348775</td>
<td align="right">1</td>
</tr>
</tbody>
</table>
<p>Note that the constant in the second format must be equal to 1. Although it seems that these two formats are different, they both have the same statistics zscore. As a matter of fact, we only require the zscore in the model rather than beta and standard error. Here, we provide these two formats to be in line with the input file of the model CoMM_S2. Similarly, the beta and se column in the GWAS summary statistics file can also be replaced with zscore and constant. At last, the corresponding file with all the information of SNPs in eQTL summary file is</p>
<table class="table">
<thead><tr class="header">
<th align="right">V1</th>
<th align="left">V2</th>
<th align="right">V3</th>
<th align="right">V4</th>
<th align="left">V5</th>
<th align="left">V6</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="left">rs201725126</td>
<td align="right">169844515</td>
<td align="right">169844515</td>
<td align="left">G</td>
<td align="left">T</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="left">rs200579949</td>
<td align="right">169794359</td>
<td align="right">169794359</td>
<td align="left">G</td>
<td align="left">A</td>
</tr>
<tr class="odd">
<td align="right">1</td>
<td align="left">rs75454623</td>
<td align="right">169766298</td>
<td align="right">169766298</td>
<td align="left">T</td>
<td align="left">G</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="left">rs199856693</td>
<td align="right">169794025</td>
<td align="right">169794025</td>
<td align="left">T</td>
<td align="left">C</td>
</tr>
<tr class="odd">
<td align="right">1</td>
<td align="left">rs78601809</td>
<td align="right">169851159</td>
<td align="right">169851159</td>
<td align="left">A</td>
<td align="left">G</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="left">rs200482301</td>
<td align="right">169759370</td>
<td align="right">169759370</td>
<td align="left">A</td>
<td align="left">G</td>
</tr>
<tr class="odd">
<td align="right">1</td>
<td align="left">rs140337953</td>
<td align="right">169921175</td>
<td align="right">169921175</td>
<td align="left">G</td>
<td align="left">A</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="left">rs200943160</td>
<td align="right">169858192</td>
<td align="right">169858192</td>
<td align="left">A</td>
<td align="left">G</td>
</tr>
<tr class="odd">
<td align="right">1</td>
<td align="left">rs116400033</td>
<td align="right">169851731</td>
<td align="right">169851731</td>
<td align="left">A</td>
<td align="left">G</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="left">rs61999471</td>
<td align="right">169860473</td>
<td align="right">169860473</td>
<td align="left">G</td>
<td align="left">A</td>
</tr>
</tbody>
</table>
<p>One point we have to explain is that this file must contain all the SNP information of eQTL summary statistics. Furthermore, to speed up model ‘TransCoMM_S2’, we implement multiple threads version of ‘TransCoMM_S2’ by just running <code>fm = TransCoMM_S2_testing_mt(stringname1, stringname2, stringname3, stringname4, stringname5, px, lam, coreNum);</code></p>
</div>
<div id="comm-a-unified-function-to-dissect-genetic-contributions-to-complex-traits-by-leveraging-regulatory-information" class="section level3">
<h3 class="hasAnchor">
<a href="#comm-a-unified-function-to-dissect-genetic-contributions-to-complex-traits-by-leveraging-regulatory-information" class="anchor"></a>CoMM: a unified function to dissect genetic contributions to complex traits by leveraging regulatory information</h3>
<p>The general form of CoMM is CoMM(Model, file1, file2, file3, file4, file5, file6, Cov1, Cov2, R1, R2, whCol, bw, lam, paral, coreNum). In fact, The function CoMM is the combination of CoMM_testing_run, CoMM_S2_testing and TransCoMM_S2_testing to facilitate the usage of these functions. Firstly, we describe the parameters used in the function CoMM in detail.</p>
<p>Model</p>
<p>The type of model specified by the user. It is a character which has four values, “Individual”, “Summary”, “Trans” and “Shared”. “Individual” stands for the model using both individual-level transcriptome and GWAS dataset. “Summary” stands for the model using individual-level transcriptome but summary-level GWAS dataset. “Trans” stands for the model using both summary-level transcriptome and GWAS dataset with different reference panels. “Shared” stands for the model using both summary-level transcriptome and GWAS dataset with the same reference panel. The default value is “Individual”.</p>
<p>file1</p>
<p>The prefix for eQTL genotype file with plink format (bim, bed, fam). The parameter file1 must be specified when the parameter Model takes value “Individual” or “Summary”. In other scenarios, the parameter file1 will be ignored.</p>
<p>file2</p>
<p>The gene expression file with full name. The parameter file2 must be specified when the parameter Model takes value “Individual” or “Summary”. But in other scenarios, the parameter file2 will be ignored. For the gene expression file, it must have the following format (rows for genes and note that it must be tab delimited):</p>
<table class="table">
<colgroup>
<col width="10%">
<col width="10%">
<col width="16%">
<col width="14%">
<col width="20%">
<col width="4%">
<col width="10%">
<col width="11%">
</colgroup>
<thead><tr class="header">
<th align="right">lower</th>
<th align="right">up</th>
<th align="left">genetype1</th>
<th align="left">genetype2</th>
<th align="left">TargetID</th>
<th align="right">Chr</th>
<th align="right">HG00105</th>
<th align="right">HG00115</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">59783540</td>
<td align="right">59843484</td>
<td align="left">lincRNA</td>
<td align="left">PART1</td>
<td align="left">ENSG00000152931.6</td>
<td align="right">5</td>
<td align="right">0.5126086</td>
<td align="right">0.7089508</td>
</tr>
<tr class="even">
<td align="right">48128225</td>
<td align="right">48148330</td>
<td align="left">protein_coding</td>
<td align="left">UPP1</td>
<td align="left">ENSG00000183696.9</td>
<td align="right">7</td>
<td align="right">1.4118007</td>
<td align="right">-0.0135644</td>
</tr>
<tr class="odd">
<td align="right">57846106</td>
<td align="right">57853063</td>
<td align="left">protein_coding</td>
<td align="left">INHBE</td>
<td align="left">ENSG00000139269.2</td>
<td align="right">12</td>
<td align="right">0.5755268</td>
<td align="right">-1.0162217</td>
</tr>
<tr class="even">
<td align="right">116054583</td>
<td align="right">116164515</td>
<td align="left">protein_coding</td>
<td align="left">AFAP1L2</td>
<td align="left">ENSG00000169129.8</td>
<td align="right">10</td>
<td align="right">1.1117776</td>
<td align="right">0.0407033</td>
</tr>
<tr class="odd">
<td align="right">22157909</td>
<td align="right">22396763</td>
<td align="left">protein_coding</td>
<td align="left">RAPGEF5</td>
<td align="left">ENSG00000136237.12</td>
<td align="right">7</td>
<td align="right">0.2831573</td>
<td align="right">-0.1772559</td>
</tr>
<tr class="even">
<td align="right">11700964</td>
<td align="right">11743303</td>
<td align="left">lincRNA</td>
<td align="left">RP11-434C1.1</td>
<td align="left">ENSG00000247157.2</td>
<td align="right">12</td>
<td align="right">0.2550282</td>
<td align="right">-0.2831573</td>
</tr>
</tbody>
</table>
<p>file3</p>
<p>The prefix for GWAS genotype and phenotype file with plink format (bim, bed, fam). The parameter file3 must be specified when the parameter Model takes value “Individual”. But in other scenarios, the parameter file3 will be ignored.</p>
<p>file4</p>
<p>The summary-level transcriptome dataset. The parameter file4 must be specified when the parameter Model takes value “Trans” or “Shared”. But in other scenarios, the parameter file4 will be ignored. For summary-level transcriptome data file, it must have the following format (note that it must be tab delimited):</p>
<table class="table">
<thead><tr class="header">
<th align="left">SNP</th>
<th align="left">Gene</th>
<th align="right">beta</th>
<th align="right">se</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">rs201725126</td>
<td align="left">ENSG00000227232</td>
<td align="right">0.6142857</td>
<td align="right">0.6780778</td>
</tr>
<tr class="even">
<td align="left">rs200579949</td>
<td align="left">ENSG00000227232</td>
<td align="right">-0.7585898</td>
<td align="right">0.0453431</td>
</tr>
<tr class="odd">
<td align="left">rs75454623</td>
<td align="left">ENSG00000227232</td>
<td align="right">0.2414931</td>
<td align="right">0.4935305</td>
</tr>
<tr class="even">
<td align="left">rs199856693</td>
<td align="left">ENSG00000170113</td>
<td align="right">0.5050623</td>
<td align="right">0.2252596</td>
</tr>
<tr class="odd">
<td align="left">rs78601809</td>
<td align="left">ENSG00000170113</td>
<td align="right">1.5409776</td>
<td align="right">0.7169221</td>
</tr>
<tr class="even">
<td align="left">rs200482301</td>
<td align="left">ENSG00000170113</td>
<td align="right">0.2524854</td>
<td align="right">0.2358414</td>
</tr>
<tr class="odd">
<td align="left">rs140337953</td>
<td align="left">ENSG00000140157</td>
<td align="right">0.5835819</td>
<td align="right">0.2434627</td>
</tr>
<tr class="even">
<td align="left">rs200943160</td>
<td align="left">ENSG00000140157</td>
<td align="right">0.0126301</td>
<td align="right">0.6551085</td>
</tr>
<tr class="odd">
<td align="left">rs116400033</td>
<td align="left">ENSG00000140157</td>
<td align="right">0.8584214</td>
<td align="right">1.7336470</td>
</tr>
<tr class="even">
<td align="left">rs61999471</td>
<td align="left">ENSG00000140157</td>
<td align="right">-0.3483619</td>
<td align="right">1.8034719</td>
</tr>
</tbody>
</table>
<p>file5</p>
<p>The prefix for the complementary summary-level transcriptome data file with suffix “.fam”, which have complete information of SNP used in file4. For example, the file can be named as “cis-eQTLs_full_chr1.fam”. The parameter file5 must be specified when the parameter Model takes value “Trans” or “Shared”. But in other scenarios, the parameter file5 will be ignored. For the complementary summary-level transcriptome data file, it must have the following format (note that it must be tab delimited):</p>
<table class="table">
<thead><tr class="header">
<th align="right">V1</th>
<th align="left">V2</th>
<th align="right">V3</th>
<th align="right">V4</th>
<th align="left">V5</th>
<th align="left">V6</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="left">rs201725126</td>
<td align="right">169844515</td>
<td align="right">169844515</td>
<td align="left">G</td>
<td align="left">T</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="left">rs200579949</td>
<td align="right">169794359</td>
<td align="right">169794359</td>
<td align="left">G</td>
<td align="left">A</td>
</tr>
<tr class="odd">
<td align="right">1</td>
<td align="left">rs75454623</td>
<td align="right">169766298</td>
<td align="right">169766298</td>
<td align="left">T</td>
<td align="left">G</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="left">rs199856693</td>
<td align="right">169794025</td>
<td align="right">169794025</td>
<td align="left">T</td>
<td align="left">C</td>
</tr>
<tr class="odd">
<td align="right">1</td>
<td align="left">rs78601809</td>
<td align="right">169851159</td>
<td align="right">169851159</td>
<td align="left">A</td>
<td align="left">G</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="left">rs200482301</td>
<td align="right">169759370</td>
<td align="right">169759370</td>
<td align="left">A</td>
<td align="left">G</td>
</tr>
<tr class="odd">
<td align="right">1</td>
<td align="left">rs140337953</td>
<td align="right">169921175</td>
<td align="right">169921175</td>
<td align="left">G</td>
<td align="left">A</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="left">rs200943160</td>
<td align="right">169858192</td>
<td align="right">169858192</td>
<td align="left">A</td>
<td align="left">G</td>
</tr>
<tr class="odd">
<td align="right">1</td>
<td align="left">rs116400033</td>
<td align="right">169851731</td>
<td align="right">169851731</td>
<td align="left">A</td>
<td align="left">G</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="left">rs61999471</td>
<td align="right">169860473</td>
<td align="right">169860473</td>
<td align="left">G</td>
<td align="left">A</td>
</tr>
</tbody>
</table>
<p>file6</p>
<p>The summary-level GWAS dataset. The parameter file6 must be specified when the parameter Model takes value “Summary”, “Trans” or “Shared”. But in other scenarios, the parameter file6 will be ignored. For the summary-level GWAS data file, it must have the following format (note that it must be tab delimited):</p>
<table class="table">
<thead><tr class="header">
<th align="left">SNP</th>
<th align="right">chr</th>
<th align="right">BP</th>
<th align="left">A1</th>
<th align="left">A2</th>
<th align="right">beta</th>
<th align="right">se</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">rs3094315</td>
<td align="right">1</td>
<td align="right">752566</td>
<td align="left">G</td>
<td align="left">A</td>
<td align="right">-0.0122</td>
<td align="right">0.0294</td>
</tr>
<tr class="even">
<td align="left">rs3128117</td>
<td align="right">1</td>
<td align="right">944564</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">-0.0208</td>
<td align="right">0.0278</td>
</tr>
<tr class="odd">
<td align="left">rs1891906</td>
<td align="right">1</td>
<td align="right">950243</td>
<td align="left">C</td>
<td align="left">A</td>
<td align="right">-0.0264</td>
<td align="right">0.0260</td>
</tr>
<tr class="even">
<td align="left">rs2710888</td>
<td align="right">1</td>
<td align="right">959842</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">-0.0439</td>
<td align="right">0.0297</td>
</tr>
<tr class="odd">
<td align="left">rs4970393</td>
<td align="right">1</td>
<td align="right">962606</td>
<td align="left">G</td>
<td align="left">A</td>
<td align="right">-0.0252</td>
<td align="right">0.0233</td>
</tr>
<tr class="even">
<td align="left">rs7526076</td>
<td align="right">1</td>
<td align="right">998395</td>
<td align="left">A</td>
<td align="left">G</td>
<td align="right">-0.0512</td>
<td align="right">0.0229</td>
</tr>
<tr class="odd">
<td align="left">rs4075116</td>
<td align="right">1</td>
<td align="right">1003629</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">-0.0497</td>
<td align="right">0.0220</td>
</tr>
<tr class="even">
<td align="left">rs3934834</td>
<td align="right">1</td>
<td align="right">1005806</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">0.0364</td>
<td align="right">0.0256</td>
</tr>
<tr class="odd">
<td align="left">rs3766192</td>
<td align="right">1</td>
<td align="right">1017197</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">-0.0116</td>
<td align="right">0.0178</td>
</tr>
<tr class="even">
<td align="left">rs3766191</td>
<td align="right">1</td>
<td align="right">1017587</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">0.0318</td>
<td align="right">0.0262</td>
</tr>
</tbody>
</table>
<p>Cov1</p>
<p>The covariates of transcriptome dataset. It is meaningful when the parameter Model takes value “Individual” or “Summary”. For other values, it can be ignored.</p>
<p>Cov2</p>
<p>The covariates of GWAS dataset. It is meaningful when the parameter Model takes value “Individual”. For other values, it can be ignored.</p>
<p>R1</p>
<p>The prefix for the reference panel corresponding to transcriptome dataset with plink format (bim, bed, fam). When the parameter Model takes value “Shared”, both the correlation matrix for eQTL and GWAS are estimated by R1. It is meaningful when the parameter Model is “Trans” or “Shared”. In other scenarios, it can be ignored.</p>
<p>R2</p>
<p>The prefix for the reference panel corresponding to GWAS dataset with plink format (bim, bed, fam). It is meaningful when the parameter Model takes value “Summary” or “Trans”. In other scenarios, it can be ignored.</p>
<p>whCol</p>
<p>Specify which column of phenotype in the fam file is adopted. For example, when whCol = 2, the 2-th column of fam file is adopted as the phenotype. The default value is 1. It is meaningful when the parameter Model takes value “Individual”. Otherwise, it will be ignored.</p>
<p>bw</p>
<p>The number of downstream and upstream SNPs that are considered as cis-SNP within a gene. The default value is 500000. This parameter is meaningful when the parameter Model takes value “Individual” or “Summary”. For other values “Trans” and “Shared”, it can be ignored.</p>
<p>lam</p>
<p>The shrinkage intensity to estimate the correlation matrix. For example, when lam = 0.95, R1 = 0.95R + 0.05I. The default value is 0.95.</p>
<p>paral</p>
<p>The logical value to specify whether the program run in parallel. The default value is F.</p>
<p>coreNum</p>
<p>The number of cores used when the program run in parallel. The default value is 1. The parameter is meaningful when the parameter paral takes logical value T. Otherwise, it will be ignored.</p>
</div>
<div id="examples" class="section level3">
<h3 class="hasAnchor">
<a href="#examples" class="anchor"></a>Examples</h3>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">file1</span> <span class="op">=</span> <span class="st">"1000G.EUR.QC"</span>;
<span class="va">file2</span> <span class="op">=</span> <span class="st">"Geuvadis_gene_expression.txt"</span>;
<span class="va">file3</span> <span class="op">=</span> <span class="st">"NFBC_filter_mph10"</span>;
<span class="va">file4</span> <span class="op">=</span> <span class="st">"cis-eQTLs_full_chr1_1-100000.txt"</span>;
<span class="va">file5</span> <span class="op">=</span> <span class="st">"cis-eQTLs_full_chr1"</span>;
<span class="va">file6</span> <span class="op">=</span> <span class="st">"NFBC_ph1_beta_se.txt"</span>;
<span class="va">R1</span> <span class="op">=</span> <span class="st">"all_chr_1000G"</span>;
<span class="va">R2</span> <span class="op">=</span> <span class="st">"NFBC_filter_mph10_400"</span>;
<span class="va">whCol</span> <span class="op">=</span> <span class="fl">1</span>;
<span class="va">bw</span> <span class="op">=</span> <span class="fl">500000</span>;
<span class="va">lam</span> <span class="op">=</span> <span class="fl">0.95</span>;
<span class="va">paral</span> <span class="op">=</span> <span class="cn">T</span>;
<span class="va">coreNum</span> <span class="op">=</span> <span class="fl">24</span>;</code></pre></div>
<p>When we want to fit CoMM using individual-level transcriptome and GWAS data, we run</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CoMM.html">CoMM</a></span><span class="op">(</span>Model <span class="op">=</span> <span class="st">"Individual"</span>, file1 <span class="op">=</span> <span class="va">file1</span>, file2 <span class="op">=</span> <span class="va">file2</span>, file3 <span class="op">=</span> <span class="va">file3</span>, whCol <span class="op">=</span> <span class="va">whCol</span>, bw <span class="op">=</span> <span class="va">bw</span><span class="op">)</span></code></pre></div>
<p>We can also run in parallel</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CoMM.html">CoMM</a></span><span class="op">(</span>Model <span class="op">=</span> <span class="st">"Individual"</span>, file1 <span class="op">=</span> <span class="va">file1</span>, file2 <span class="op">=</span> <span class="va">file2</span>, file3 <span class="op">=</span> <span class="va">file3</span>, whCol <span class="op">=</span> <span class="va">whCol</span>, bw <span class="op">=</span> <span class="va">bw</span>, paral <span class="op">=</span> <span class="va">paral</span>, coreNum <span class="op">=</span> <span class="va">coreNum</span><span class="op">)</span></code></pre></div>
<p>When we can only access to the individual-level transcriptome dataset and the summary-level GWAS dataset. In addition, we have the dataset “NFBC_filter_mph10_400” as the reference panel. Then we can run</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CoMM.html">CoMM</a></span><span class="op">(</span>Model <span class="op">=</span> <span class="st">"Summary"</span>, file1 <span class="op">=</span> <span class="va">file1</span>, file2 <span class="op">=</span> <span class="va">file2</span>, file6 <span class="op">=</span> <span class="va">file6</span>, R2 <span class="op">=</span> <span class="va">R2</span>, bw <span class="op">=</span><span class="va">bw</span><span class="op">)</span></code></pre></div>
<p>We can also run in parallel</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CoMM.html">CoMM</a></span><span class="op">(</span>Model <span class="op">=</span> <span class="st">"Summary"</span>, file1 <span class="op">=</span> <span class="va">file1</span>, file2 <span class="op">=</span> <span class="va">file2</span>, file6 <span class="op">=</span> <span class="va">file6</span>, R2 <span class="op">=</span> <span class="va">R2</span>, bw <span class="op">=</span><span class="va">bw</span>, paral <span class="op">=</span> <span class="va">paral</span>, coreNum <span class="op">=</span> <span class="va">coreNum</span><span class="op">)</span></code></pre></div>
<p>Moreover, we can also specify the shrinkage intensity</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CoMM.html">CoMM</a></span><span class="op">(</span>Model <span class="op">=</span> <span class="st">"Summary"</span>, file1 <span class="op">=</span> <span class="va">file1</span>, file2 <span class="op">=</span> <span class="va">file2</span>, file6 <span class="op">=</span> <span class="va">file6</span>, R2 <span class="op">=</span> <span class="va">R2</span>, bw <span class="op">=</span><span class="va">bw</span>, lam <span class="op">=</span> <span class="va">lam</span>, paral <span class="op">=</span> <span class="va">paral</span>, coreNum <span class="op">=</span> <span class="va">coreNum</span><span class="op">)</span></code></pre></div>
<p>When we both have the summary-level transcriptome dataset and GWAS dataset. In addition, we adopt the binary plink files “all_chr_1000G” and “NFBC_filter_mph10_400” as the reference panel for transcriptome and GWAS datasets, respectively</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CoMM.html">CoMM</a></span><span class="op">(</span>Model <span class="op">=</span> <span class="st">"Trans"</span>, file4 <span class="op">=</span> <span class="va">file4</span>, file5 <span class="op">=</span> <span class="va">file5</span>, file6 <span class="op">=</span> <span class="va">file6</span>, R1 <span class="op">=</span> <span class="va">R1</span>, R2 <span class="op">=</span> <span class="va">R2</span><span class="op">)</span></code></pre></div>
<p>In this setting, we can also specify the shrinkage intensity and run in parallel</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CoMM.html">CoMM</a></span><span class="op">(</span>Model <span class="op">=</span> <span class="st">"Trans"</span>, file4 <span class="op">=</span> <span class="va">file4</span>, file5 <span class="op">=</span> <span class="va">file5</span>, file6 <span class="op">=</span> <span class="va">file6</span>, R1 <span class="op">=</span> <span class="va">R1</span>, R2 <span class="op">=</span> <span class="va">R2</span>, lam <span class="op">=</span> <span class="va">lam</span>, paral <span class="op">=</span> <span class="va">paral</span>, coreNum <span class="op">=</span> <span class="va">coreNum</span><span class="op">)</span></code></pre></div>
<p>When we both have the summary-level transcriptome dataset and GWAS dataset which share the same dataset “all_chr_1000G” as the reference panel. We run</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CoMM.html">CoMM</a></span><span class="op">(</span>Model <span class="op">=</span> <span class="st">"Shared"</span>, file4 <span class="op">=</span> <span class="va">file4</span>, file5 <span class="op">=</span> <span class="va">file5</span>, file6 <span class="op">=</span> <span class="va">file6</span>, R1 <span class="op">=</span> <span class="va">R1</span><span class="op">)</span></code></pre></div>
<p>Similarly, we we can specify the shrinkage intensity and run in parallel</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CoMM.html">CoMM</a></span><span class="op">(</span>Model <span class="op">=</span> <span class="st">"Shared"</span>, file4 <span class="op">=</span> <span class="va">file4</span>, file5 <span class="op">=</span> <span class="va">file5</span>, file6 <span class="op">=</span> <span class="va">file6</span>, R1 <span class="op">=</span> <span class="va">R1</span>, lam <span class="op">=</span> <span class="va">lam</span>, paral <span class="op">=</span> <span class="va">paral</span>, coreNum <span class="op">=</span> <span class="va">coreNum</span><span class="op">)</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jin Liu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
