# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

CoMM_S2_paral_testing <- function(stringname1, stringname2, stringname3, stringname4, stringname5, bw = 500000L, lam = 0.95, coreNum = 1L) {
    .Call('_CoMM_CoMM_S2_paral_testing', PACKAGE = 'CoMM', stringname1, stringname2, stringname3, stringname4, stringname5, bw, lam, coreNum)
}

CoMM_S2_testing <- function(stringname1, stringname2, stringname3, stringname4, stringname5, bw = 500000L, lam = 0.95, alpha = 0) {
    .Call('_CoMM_CoMM_S2_testing', PACKAGE = 'CoMM', stringname1, stringname2, stringname3, stringname4, stringname5, bw, lam, alpha)
}

CoMM_S4 <- function(hatmur, hatmu2r, Rr, R2r, opts = NULL, px = 1L) {
    .Call('_CoMM_CoMM_S4', PACKAGE = 'CoMM', hatmur, hatmu2r, Rr, R2r, opts, px)
}

read_eqtls <- function(filename, P) {
    .Call('_CoMM_read_eqtls', PACKAGE = 'CoMM', filename, P)
}

CoMM_S4_testing <- function(stringname1, stringname2, stringname3, stringname4, stringname5, px = 1L, lam = 0.95) {
    .Call('_CoMM_CoMM_S4_testing', PACKAGE = 'CoMM', stringname1, stringname2, stringname3, stringname4, stringname5, px, lam)
}

CoMM_S4_testing_mt <- function(stringname1, stringname2, stringname3, stringname4, stringname5, px = 1L, lam = 0.95, coreNum = 24L) {
    .Call('_CoMM_CoMM_S4_testing_mt', PACKAGE = 'CoMM', stringname1, stringname2, stringname3, stringname4, stringname5, px, lam, coreNum)
}

#' @author Jin Liu, \email{jin.liu@duke-nus.edu.sg}
#' @title
#' CoMM
#' @description
#' CoMM to dissecting genetic contributions to complex traits by leveraging regulatory information.
#'
#' @param y  gene expression vector.
#' @param z  trait vector.
#' @param x1  normalized genotype (cis-SNPs) matrix for eQTL.
#' @param x2  normalized genotype (cis-SNPs) matrix for GWAS.
#' @param w1  covariates file for eQTL data.
#' @param w2  covariates file for GWAS data, e.g. top 10 PCs.
#' @param constr  indicator for constraint (alpha = 0 if constr = 1).
#' @param epsStopLogLik  convergence criteria (default is 1e-5).
#' @param maxIter  maximum iteration (default is 1000).
#' @param pxem_indicator  indicator for using PX-EM (default is 1). pxem_indicator = 1 for using PX-EM and EM otherwise.
#'
#' @return List of model parameters
#'
#' @examples
#' L = 1;
#' M = 100;
#' rho =0.5;
#' n1 = 350;
#' n2 = 5000;
#'
#' X <- matrix(rnorm((n1+n2)*M),nrow=n1+n2,ncol=M);
#' beta_prop = 0.2;
#' b = numeric(); m = M * beta_prop;
#' b[sample(M,m)] = rnorm(m); h2y = 0.05;
#' y0 <- X%*%b + 6;
#' y  <- y0 + (as.vector(var(y0)*(1-h2y)/h2y))^0.5*rnorm(n1+n2);
#'
#' h2 = 0.001;
#' y1 <- y[1:n1]
#' X1 <- X[1:n1,]
#' y2 <- y0[(n1+1):(n1+n2)]
#' X2 <- X[(n1+1):(n1+n2),]
#' alpha0 <- 3
#' alpha <- 0.3
#' sz2 <- var(y2*alpha) * ((1-h2)/h2)
#' z <- alpha0 + y2*alpha + rnorm(n2,0,sqrt(sz2))
#' y = y1;
#'
#' mean.x1 = apply(X1,2,mean);
#' x1m = sweep(X1,2,mean.x1);
#' std.x1 = apply(x1m,2,sd)
#' x1p = sweep(x1m,2,std.x1,"/");
#' x1p = x1p/sqrt(dim(x1p)[2])
#'
#'  mean.x2 = apply(X2,2,mean);
#' x2m = sweep(X2,2,mean.x2);
#' std.x2 = apply(x2m,2,sd)
#' x2p = sweep(x2m,2,std.x2,"/");
#' x2p = x2p/sqrt(dim(x2p)[2])
#'
#' w2 = matrix(rep(1,n2),ncol=1);
#' w1 = matrix(rep(1,n1),ncol=1);
#' fm = CoMM_covar_pxem(y,z,x1p,x2p,w1,w2);
#'
#' @details
#' \code{CoMM_covar_pxem} fits the CoMM model using raw data.
#' @export
CoMM_covar_pxem <- function(y, z, x1, x2, w1, w2, constr = 0L, epsStopLogLik = 1e-5, maxIter = 1000L, pxem_indicator = 1L) {
    .Call('_CoMM_CoMM_covar_pxem', PACKAGE = 'CoMM', y, z, x1, x2, w1, w2, constr, epsStopLogLik, maxIter, pxem_indicator)
}

#' @author Jin Liu, \email{jin.liu@duke-nus.edu.sg}
#' @title
#' CoMM
#' @description
#' CoMM to dissecting genetic contributions to complex traits by leveraging regulatory information.
#'
#' @param stringname1  prefix for eQTL genotype file with plink format (bim, bed).
#' @param stringname2  prefix for GWAS genotype and phenotype file with plink format (bim, bed, fam).
#' @param stringname3  gene expression file with full name.
#' @param stringname4  covariates file for eQTL data.
#' @param stringname5  covariates file for GWAS data, e.g. top 10 PCs.
#' @param whCol  specify which phenotype is used in fam. For example, when whCol = 2, the seven-th column of fam file will be used as phenotype.
#' @param bw  the number of downstream and upstream SNPs that are considered as cis-SNP within a gene.
#' @param coreNum the number of core for parallelization. 
#'
#' @return List of model parameters
#'
#' @examples
#' file1 = "1000G.EUR.QC.1";
#' file2 = "NFBC_filter_mph10";
#' file3 = "Geuvadis_gene_expression_qn.txt";
#' file4 = "";
#' file5 = "pc5_NFBC_filter_mph10.txt";
#' whichPheno = 1;
#' bw = 500000;
#' coreNum = 24;
#'
#' fm = CoMM_testing_run_mt(file1,file2,file3, file4,file5, whichPheno, bw, coreNum);
#'
#' @details
#' \code{CoMM_testing_run_mt} fits the CoMM model. It requires to provide plink binary eQTL genotype file (bim, bed)
#' the GWAS plink binary file (bim, bed, fam), gene expression file for eQTL.
#' @export
CoMM_testing_run_mt <- function(stringname1, stringname2, stringname3, stringname4, stringname5, whCol, bw, coreNum) {
    .Call('_CoMM_CoMM_testing_run_mt', PACKAGE = 'CoMM', stringname1, stringname2, stringname3, stringname4, stringname5, whCol, bw, coreNum)
}

#' @author Jin Liu, \email{jin.liu@duke-nus.edu.sg}
#' @title
#' CoMM
#' @description
#' CoMM to dissecting genetic contributions to complex traits by leveraging regulatory information.
#'
#' @param stringname1  prefix for eQTL genotype file with plink format (bim, bed).
#' @param stringname2  prefix for GWAS genotype and phenotype file with plink format (bim, bed, fam).
#' @param stringname3  gene expression file with full name.
#' @param stringname4  covariates file for eQTL data.
#' @param stringname5  covariates file for GWAS data, e.g. top 10 PCs.
#' @param whCol  specify which phenotype is used in fam. For example, when whCol = 2, the seven-th column of fam file will be used as phenotype.
#' @param bw  the number of downstream and upstream SNPs that are considered as cis-SNP within a gene.
#'
#' @return List of model parameters
#'
#' @examples
#' ##Working with no summary statistics, no covariates and options
#' file1 = "1000G.EUR.QC.1";
#' file2 = "NFBC_filter_mph10";
#' file3 = "Geuvadis_gene_expression_qn.txt";
#' file4 = "";
#' file5 = "pc5_NFBC_filter_mph10.txt";
#' whichPheno = 1;
#' bw = 500000;
#'
#' fm = CoMM_testing_run(file1,file2,file3, file4,file5, whichPheno, bw);
#'
#' @details
#' \code{CoMM_testing_run} fits the CoMM model. It requires to provide plink binary eQTL genotype file (bim, bed)
#' the GWAS plink binary file (bim, bed, fam), gene expression file for eQTL.
#' @export
CoMM_testing_run <- function(stringname1, stringname2, stringname3, stringname4, stringname5, whCol, bw) {
    .Call('_CoMM_CoMM_testing_run', PACKAGE = 'CoMM', stringname1, stringname2, stringname3, stringname4, stringname5, whCol, bw)
}

CoMM_S2 <- function(xr, yr, Wr, hatmur, hatsr, Rr, opts = NULL, px = 1L) {
    .Call('_CoMM_CoMM_S2', PACKAGE = 'CoMM', xr, yr, Wr, hatmur, hatsr, Rr, opts, px)
}

#' @author Jin Liu, \email{jin.liu@duke-nus.edu.sg}
#' @title
#' CoMM
#' @description
#' CoMM to dissecting genetic contributions to complex traits by leveraging regulatory information.
#'
#' @param y  gene expression vector.
#' @param x  normalized genotype (cis-SNPs) matrix for eQTL.
#' @param w  covariates file for eQTL data.
#' @param maxIter  maximum iteration (default is 1000).
#'
#' @return List of model parameters
#'
#' @examples
#' L = 1; M = 100; rho =0.5
#' n1 = 350; n2 = 5000;
#' X <- matrix(rnorm((n1+n2)*M),nrow=n1+n2,ncol=M);
#'
#' beta_prop = 0.2;
#' b = numeric(M);
#' m = M * beta_prop;
#' b[sample(M,m)] = rnorm(m);
#' h2y = 0.05;
#' b0 = 6;
#'
#' y0 <- X%*%b + b0;
#' y  <- y0 + (as.vector(var(y0)*(1-h2y)/h2y))^0.5*rnorm(n1+n2);
#' h2 = 0.001;
#' y1 <- y[1:n1]
#' X1 <- X[1:n1,]
#' y = y1;
#'
#' mean.x1 = apply(X1,2,mean);
#' x1m = sweep(X1,2,mean.x1);
#' std.x1 = apply(x1m,2,sd)
#' x1p = sweep(x1m,2,std.x1,"/");
#' x1p = x1p/sqrt(dim(x1p)[2])
#' w1 = matrix(rep(1,n1),ncol=1);
#'
#' fm0 = lmm_pxem(y, w1,x1p, 100)
#'
#' @details
#' \code{lmm_pxem} fits the linear mixed model. (n < p)
#' @export
lmm_pxem <- function(y, w, x, maxIter) {
    .Call('_CoMM_lmm_pxem', PACKAGE = 'CoMM', y, w, x, maxIter)
}

lmm_pxem_test <- function(y, w, x, maxIter) {
    .Call('_CoMM_lmm_pxem_test', PACKAGE = 'CoMM', y, w, x, maxIter)
}

#' @author Jin Liu, \email{jin.liu@duke-nus.edu.sg}
#' @title
#' CoMM
#' @description
#' CoMM to dissecting genetic contributions to complex traits by leveraging regulatory information.
#'
#' @param y  gene expression vector.
#' @param x  normalized genotype (cis-SNPs) matrix for eQTL.
#' @param w  covariates file for eQTL data.
#' @param maxIter  maximum iteration (default is 1000).
#'
#' @return List of model parameters
#'
#' @examples
#' L = 1; M = 100; rho =0.5
#' n1 = 350; n2 = 5000;
#' X <- matrix(rnorm((n1+n2)*M),nrow=n1+n2,ncol=M);
#'
#' beta_prop = 0.2;
#' b = numeric(M);
#' m = M * beta_prop;
#' b[sample(M,m)] = rnorm(m);
#' h2y = 0.05;
#' b0 = 6;
#'
#' y0 <- X%*%b + b0;
#' y  <- y0 + (as.vector(var(y0)*(1-h2y)/h2y))^0.5*rnorm(n1+n2);
#' h2 = 0.001;
#' y1 <- y[1:n1]
#' X1 <- X[1:n1,]
#' y = y1;
#'
#' mean.x1 = apply(X1,2,mean);
#' x1m = sweep(X1,2,mean.x1);
#' std.x1 = apply(x1m,2,sd)
#' x1p = sweep(x1m,2,std.x1,"/");
#' x1p = x1p/sqrt(dim(x1p)[2])
#' w1 = matrix(rep(1,n1),ncol=1);
#'
#' fm0 = lmm_pxem2(y, w1,x1p, 100)
#'
#' @details
#' \code{lmm_pxem2} fits the linear mixed model (n > p).
#' @export
lmm_pxem2 <- function(y, w, x, maxIter) {
    .Call('_CoMM_lmm_pxem2', PACKAGE = 'CoMM', y, w, x, maxIter)
}

